geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoVisibilityRowEncoder.scala

```diff
@@ -26,52 +27,117 @@ import org.opengis.feature.simple.SimpleFeatureType
 class KryoVisibilityRowEncoder extends RowEncodingIterator {
 
   private var sft: SimpleFeatureType = _
-  private var nullBytes: Array[Array[Byte]] = _
-  private var offsets: Array[Int] = _
+  private var count: Int = -1
+  private var attributes: Array[(Array[Byte], Int, Int)] = _ // (bytes, offset, length)
+
+  private var nullBytesV2: Array[Array[Byte]] = _
+  private var offsetsV2: Array[Int] = _
   private var offsetStart: Int = -1
 
   private val output: Output = new Output(128, -1)
 
-  override def init(source: SortedKeyValueIterator[Key, Value],
-                    options: java.util.Map[String, String],
-                    env: IteratorEnvironment): Unit = {
+  override def init(
+      source: SortedKeyValueIterator[Key, Value],
+      options: java.util.Map[String, String],
+      env: IteratorEnvironment): Unit = {
     super.init(source, options, env)
-
     sft = IteratorCache.sft(options.get(KryoVisibilityRowEncoder.SftOpt))
-    if (offsets == null || offsets.length != sft.getAttributeCount) {
-      offsets = Array.ofDim[Int](sft.getAttributeCount)
-    }
-    val cacheKey = CacheKeyGenerator.cacheKey(sft)
-    nullBytes = KryoFeatureSerialization.getWriters(cacheKey, sft).map { writer =>
-      output.clear()
-      writer(output, null)
-      output.toBytes
+    count = sft.getAttributeCount
+    if (attributes == null || attributes.length < count) {
+      attributes = Array.ofDim[(Array[Byte], Int, Int)](count)
     }
+    nullBytesV2 = null // lazily initialized when we hit v2 encoded data
   }
 
   override def rowEncoder(keys: java.util.List[Key], values: java.util.List[Value]): Value = {
     if (values.size() == 1) {
       return values.get(0)
     }
+    // TODO if we don't have a geometry, skip the record?
+    values.get(0).get.head match {
+      case KryoFeatureSerializer.Version  => encodeV3(keys, values)
+      case KryoFeatureSerializer.Version2 => encodeV2(keys, values)
+    }
+  }
+
+  private def encodeV3(keys: java.util.List[Key], values: java.util.List[Value]): Value = {
```

Comment:
Could you add a comment to describe the v3 and v2 formats?

Reply:
good call, thanks. I had comments in-line in the code, but capturing the overall layout is useful. I added it in the scaladocs for `KryoFeatureSerializer` and `KryoFeatureDeserializationV2`